mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,691 @@
+        """ Removes specified entities in a processed query. If all entities are being
+        removed, this function will not remove entities that the annotator does not support
+        unless it is explicitly specified to do so in the config with the param
+        "unannotate_supported_entities_only" (boolean).
+
+        Args:
+            processed_query (ProcessedQuery): A processed query.
+            remove_entities (list): List of entities to remove.
+        """
+        remove_supported_only = self.config["unannotate_supported_entities_only"]
+        keep_entities = []
+        for query_entity in processed_query.entities:
+            if remove_entities == ["*"]:
+                is_supported_entity = self.valid_entity_check(query_entity.entity.type)
+                if remove_supported_only and not is_supported_entity:
+                    keep_entities.append(query_entity)
+            elif query_entity.entity.type not in remove_entities:
+                keep_entities.append(query_entity)
+        processed_query.entities = tuple(keep_entities)
+
+    @abstractmethod
+    def parse(self, sentence, **kwargs):
+        """ Extract entities from a sentence. Detected entities should be
+        represented as dictionaries with the following keys: "body", "start"
+        (start index), "end" (end index), "value", "dim" (entity type).
+
+        Args:
+            sentence (str): Sentence to detect entities.
+
+        Returns:
+            entities (list): List of entity dictionaries.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class SpacyAnnotator(Annotator):
+    """ Annotator class that uses spacy to generate annotations.
+    """
+
+    def __init__(self, app_path, config=None):
+        super().__init__(app_path=app_path, config=config)
+
+        self.model = self.config.get("spacy_model", EN_CORE_WEB_LG)
+        self.nlp = SpacyAnnotator._load_model(self.model)
+        self.duckling = DucklingRecognizer.get_instance()
+        self.ANNOTATOR_TO_DUCKLING_ENTITY_MAPPINGS = {
+            "money": "sys_amount-of-money",
+            "cardinal": "sys_number",
+            "ordinal": "sys_ordinal",
+            "person": "sys_person",
+            "percent": "sys_percent",
+            "distance": "sys_distance",
+            "quantity": "sys_weight",
+        }
+
+    @staticmethod
+    def _load_model(model):
+        """ Load Spacy English model. Download if needed.
+
+        Args:
+            model (str): Spacy model ("en_core_web_sm", "en_core_web_md", or
+                "en_core_web_lg").
+
+        Returns:
+            nlp (spacy.lang.en.English): Spacy language model.
+        """
+
+        if model in [EN_CORE_WEB_SM, EN_CORE_WEB_MD, EN_CORE_WEB_LG]:
+            logger.info("Loading Spacy model %s.", model)
+            try:
+                return spacy.load(model)
+            except OSError:
+                os.system("python -m spacy download " + model)
```

Comment:
we should log a warning here

Reply:
Added warning

`logger.warning("%s not found. Downloading the model.", model)`

Reply:
Lets use `subprocess ` instead of `os.sysem` since thats what we use for other system calls in MindMeld. It also is non-blocking